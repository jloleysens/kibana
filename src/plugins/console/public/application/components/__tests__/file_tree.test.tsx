/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import '../../models/legacy_core_editor/legacy_core_editor.test.mocks';
// @ts-ignore
import { findTestSubject } from '@elastic/eui/lib/test';
import React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import { FileTreeEntry } from '../../components/file_tree/file_tree_entry';

import { serviceContextMock } from '../../contexts/services_context.mock';
import { TextObjectsContextProvider, ServicesContextProvider, ContextValue } from '../../contexts';

import { FileTree } from '../file_tree';

import { createHelpers } from './helpers';

describe('File Tree', () => {
  let serviceContext: ContextValue;
  let component: ReactWrapper;
  let helpers: ReturnType<typeof createHelpers>;

  beforeEach(() => {
    serviceContext = serviceContextMock.create();
    component = doMount();
    helpers = createHelpers(component);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  const doMount = () =>
    mount(
      <ServicesContextProvider value={serviceContext}>
        <TextObjectsContextProvider>
          <FileTree />
        </TextObjectsContextProvider>
      </ServicesContextProvider>
    );

  it('renders', () => {
    // Empty file tree
    expect(component).toMatchSnapshot();
  });

  const DEFAULT_CREATE_ATTRIBUTES = {
    id: 'test',
    name: 'test',
    createdAt: 123,
    updatedAt: 123,
  };

  it('creates a new file', async () => {
    const { createTestFile, fileEntryExists } = helpers;
    const name1 = 'test';
    const name2 = 'anotherTest';
    (serviceContext.services.objectStorageClient.text.create as jest.Mock)
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name1,
        name: name1,
      })
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name2,
        name: name2,
      });

    expect(component.find(FileTreeEntry).length).toBe(0);
    await createTestFile();

    component.update();
    expect(component.find(FileTreeEntry).length).toBe(1);
    expect(fileEntryExists(name1)).toBe(true);
    expect(fileEntryExists(name2)).toBe(false);

    await createTestFile();

    component.update();
    expect(component.find(FileTreeEntry).length).toBe(2);
    expect(fileEntryExists(name2)).toBe(true);
  });

  it('edits a file name', async () => {
    const { createTestFile, editFile, getFileEntry } = helpers;
    const oldName = 'test';
    const newName = 'test2';
    (serviceContext.services.objectStorageClient.text.create as jest.Mock).mockResolvedValue({
      ...DEFAULT_CREATE_ATTRIBUTES,
      id: 'test',
      name: oldName,
    });
    (serviceContext.services.objectStorageClient.text.update as jest.Mock).mockResolvedValue({
      ...DEFAULT_CREATE_ATTRIBUTES,
      id: 'test',
      name: newName,
    });

    expect(component.find(FileTreeEntry).length).toBe(0);

    await createTestFile();
    component.update();

    await editFile(oldName, newName);
    component.update();

    expect(component.find(FileTreeEntry).length).toBe(1);
    expect(getFileEntry(newName).text()).toBe(newName);
  });

  it('deletes a file', async () => {
    const { createTestFile, deleteFile, fileEntryExists } = helpers;
    const name1 = 'test';
    const name2 = 'scratchpad'; // This name is overridden with 'Default'
    expect(findTestSubject(component, `consoleModalDeleteFileButton-${name1}`).exists()).toBe(
      false
    );
    (serviceContext.services.objectStorageClient.text.create as jest.Mock)
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name1,
        name: name1,
      })
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name2,
        name: name2,
        isScratchPad: true,
      });

    await createTestFile();
    await createTestFile();

    component.update();
    expect(findTestSubject(component, `consoleDeleteFileButton-${name1}`).exists()).toBe(true);
    // The default text object should not have a delete button!
    expect(findTestSubject(component, `consoleDeleteFileButton-${name2}`).exists()).toBe(false);

    await deleteFile(name1);
    component.update();

    expect(component.find(FileTreeEntry).length).toBe(1);
    expect(fileEntryExists(name1)).toBe(false);
    expect(fileEntryExists('Default')).toBe(true);
  });

  it('shows errors', async () => {
    const { createTestFile, editFile, getFileErrorIcon } = helpers;
    const name1 = 'test';
    const newName1 = 'testAgain';
    const name2 = 'anotherTest';
    (serviceContext.services.objectStorageClient.text.create as jest.Mock)
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name1,
        name: name1,
      })
      .mockResolvedValueOnce({
        ...DEFAULT_CREATE_ATTRIBUTES,
        id: name2,
        name: name2,
        isScratchPad: true,
      });

    (serviceContext.services.objectStorageClient.text.update as jest.Mock).mockRejectedValueOnce(
      new Error('something is wrong!')
    );

    await createTestFile();
    await createTestFile();
    component.update();
    await editFile(name1, newName1);
    component.update();

    expect(getFileErrorIcon(newName1).exists()).toBe(true);
    expect(getFileErrorIcon(name1).exists()).toBe(false);
    expect(getFileErrorIcon(name2).exists()).toBe(false);
  });
});
