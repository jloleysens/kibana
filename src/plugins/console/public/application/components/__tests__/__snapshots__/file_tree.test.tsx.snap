// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Tree renders 1`] = `
<ServicesContextProvider
  value={
    Object {
      "docLinkVersion": "NA",
      "elasticsearchUrl": "test",
      "services": Object {
        "history": HistoryMock {
          "addToHistory": [MockFunction],
          "change": [MockFunction],
          "changeEmitter": BehaviorSubject {
            "_isScalar": false,
            "_value": Array [],
            "closed": false,
            "hasError": false,
            "isStopped": false,
            "observers": Array [],
            "thrownError": null,
          },
          "clearHistory": [MockFunction],
          "deleteLegacySavedEditorState": [MockFunction],
          "getHistory": [MockFunction],
          "getHistoryKeys": [MockFunction],
          "getLegacySavedEditorState": [MockFunction],
          "storage": StorageMock {
            "decode": [MockFunction],
            "decodeKey": [MockFunction],
            "delete": [MockFunction],
            "encode": [MockFunction],
            "encodeKey": [MockFunction],
            "engine": Object {},
            "get": [MockFunction],
            "has": [MockFunction],
            "keys": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Array [],
                },
              ],
            },
            "prefix": "test",
            "set": [MockFunction],
          },
          "updateCurrentState": [MockFunction],
        },
        "notifications": Object {
          "toasts": Object {
            "add": [MockFunction],
            "addDanger": [MockFunction],
            "addError": [MockFunction],
            "addSuccess": [MockFunction],
            "addWarning": [MockFunction],
            "get$": [MockFunction],
            "remove": [MockFunction],
          },
        },
        "objectStorageClient": Object {
          "text": Object {
            "create": [MockFunction],
            "delete": [MockFunction],
            "findAll": [MockFunction],
            "get": [MockFunction],
            "update": [MockFunction],
          },
        },
        "settings": undefined,
        "storage": StorageMock {
          "decode": [MockFunction],
          "decodeKey": [MockFunction],
          "delete": [MockFunction],
          "encode": [MockFunction],
          "encodeKey": [MockFunction],
          "engine": Object {},
          "get": [MockFunction],
          "has": [MockFunction],
          "keys": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Array [],
              },
            ],
          },
          "prefix": "test",
          "set": [MockFunction],
        },
        "trackUiMetric": Object {
          "count": [MockFunction],
          "load": [MockFunction],
        },
      },
    }
  }
>
  <TextObjectsContextProvider>
    <FileTree>
      <FileTree
        disabled={false}
        entries={Array []}
        onCreate={[Function]}
        onSearchFilter={[Function]}
      >
        <div
          className="conApp__fileTree"
        >
          <EuiFlexGroup
            className="conApp__fileTree__entryContainer"
            direction="column"
            gutterSize="none"
            responsive={false}
          >
            <div
              className="euiFlexGroup euiFlexGroup--directionColumn conApp__fileTree__entryContainer"
            >
              <EuiFlexItem
                grow={false}
              >
                <div
                  className="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <FileActionsBar
                    disabled={false}
                    onCreate={[Function]}
                    onFilter={[Function]}
                  >
                    <EuiFlexGroup
                      alignItems="center"
                      className="conApp__fileTree__actionBar"
                      gutterSize="none"
                      justifyContent="center"
                    >
                      <div
                        className="euiFlexGroup euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--responsive conApp__fileTree__actionBar"
                      >
                        <EuiFlexItem>
                          <div
                            className="euiFlexItem"
                          />
                        </EuiFlexItem>
                        <EuiFlexItem
                          grow={false}
                        >
                          <div
                            className="euiFlexItem euiFlexItem--flexGrowZero"
                          >
                            <EuiButtonIcon
                              aria-label="Toggle file filter"
                              color="text"
                              disabled={false}
                              iconType="search"
                              onClick={[Function]}
                              size="s"
                            >
                              <button
                                aria-label="Toggle file filter"
                                className="euiButtonIcon euiButtonIcon--text"
                                disabled={false}
                                onClick={[Function]}
                                size="s"
                                type="button"
                              >
                                <EuiIcon
                                  aria-hidden="true"
                                  className="euiButtonIcon__icon"
                                  size="m"
                                  type="search"
                                >
                                  <EuiIconEmpty
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    />
                                  </EuiIconEmpty>
                                </EuiIcon>
                              </button>
                            </EuiButtonIcon>
                          </div>
                        </EuiFlexItem>
                        <EuiFlexItem
                          grow={false}
                        >
                          <div
                            className="euiFlexItem euiFlexItem--flexGrowZero"
                          >
                            <EuiPopover
                              anchorPosition="downCenter"
                              button={
                                <EuiButtonIcon
                                  aria-label="Create a file"
                                  color="text"
                                  data-test-subj="consoleCreateFileButton"
                                  disabled={false}
                                  iconType="plusInCircle"
                                  onClick={[Function]}
                                  size="s"
                                />
                              }
                              closePopover={[Function]}
                              display="inlineBlock"
                              hasArrow={true}
                              initialFocus={[Function]}
                              isOpen={false}
                              ownFocus={true}
                              panelPaddingSize="m"
                            >
                              <EuiOutsideClickDetector
                                isDisabled={true}
                                onOutsideClick={[Function]}
                              >
                                <div
                                  className="euiPopover euiPopover--anchorDownCenter"
                                  onKeyDown={[Function]}
                                  onMouseDown={[Function]}
                                  onMouseUp={[Function]}
                                  onTouchEnd={[Function]}
                                  onTouchStart={[Function]}
                                >
                                  <div
                                    className="euiPopover__anchor"
                                  >
                                    <EuiButtonIcon
                                      aria-label="Create a file"
                                      color="text"
                                      data-test-subj="consoleCreateFileButton"
                                      disabled={false}
                                      iconType="plusInCircle"
                                      onClick={[Function]}
                                      size="s"
                                    >
                                      <button
                                        aria-label="Create a file"
                                        className="euiButtonIcon euiButtonIcon--text"
                                        data-test-subj="consoleCreateFileButton"
                                        disabled={false}
                                        onClick={[Function]}
                                        size="s"
                                        type="button"
                                      >
                                        <EuiIcon
                                          aria-hidden="true"
                                          className="euiButtonIcon__icon"
                                          size="m"
                                          type="plusInCircle"
                                        >
                                          <EuiIconEmpty
                                            aria-hidden={true}
                                            className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                                            focusable="false"
                                            role="img"
                                            style={null}
                                          >
                                            <svg
                                              aria-hidden={true}
                                              className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                                              focusable="false"
                                              height={16}
                                              role="img"
                                              style={null}
                                              viewBox="0 0 16 16"
                                              width={16}
                                              xmlns="http://www.w3.org/2000/svg"
                                            />
                                          </EuiIconEmpty>
                                        </EuiIcon>
                                      </button>
                                    </EuiButtonIcon>
                                  </div>
                                </div>
                              </EuiOutsideClickDetector>
                            </EuiPopover>
                          </div>
                        </EuiFlexItem>
                      </div>
                    </EuiFlexGroup>
                  </FileActionsBar>
                </div>
              </EuiFlexItem>
            </div>
          </EuiFlexGroup>
        </div>
      </FileTree>
    </FileTree>
  </TextObjectsContextProvider>
</ServicesContextProvider>
`;
